idtools/idmap.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/containerd/containerd/snapshots"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+const invalidID = 1<<32 - 1
+
+var invalidUser = User{Uid: invalidID, Gid: invalidID}
+
+// User is a Uid and Gid pair of a user
+//
+//nolint:revive
+type User struct {
+	Uid uint32
+	Gid uint32
+}
+
+// IDMap contains the mappings of Uids and Gids.
+//
+//nolint:revive
+type IDMap struct {
+	UidMap []specs.LinuxIDMapping `json:"UidMap"`
+	GidMap []specs.LinuxIDMapping `json:"GidMap"`
+}
+
+func LoadIDMap(id string, labels map[string]string) (IDMap, error) {
+	var idmap IDMap
+	uidmapJSON, okUID := labels[snapshots.LabelSnapshotUIDMapping]
+	gidmapJSON, okGID := labels[snapshots.LabelSnapshotGIDMapping]
+	if okUID && okGID {
+		if err := json.Unmarshal([]byte(uidmapJSON), &idmap.UidMap); err != nil {
+			return IDMap{}, err
+		}
+		if err := json.Unmarshal([]byte(gidmapJSON), &idmap.GidMap); err != nil {
+			return IDMap{}, err
+		}
+	}
+	return idmap, nil
+}
+
+// ToHost returns the mapped host user ID pair
+func (i IDMap) ToHost(pair User) (User, error) {
+	var (
+		target User
+		err    error
+	)
+	target.Uid, err = toHost(pair.Uid, i.UidMap)
+	if err != nil {
+		return invalidUser, err
+	}
+	target.Gid, err = toHost(pair.Gid, i.GidMap)
+	if err != nil {
+		return invalidUser, err
+	}
+	return target, nil
+}
+
+// Empty returns true if there are no id mappings
+func (i IDMap) Empty() bool {
+	return len(i.UidMap) == 0 && len(i.GidMap) == 0
+}
+
+// toHost takes an id mapping and a remapped ID, and translates the
+// ID to the mapped host ID. If no map is provided, then the translation
+// assumes a 1-to-1 mapping and returns the passed in id #
+func toHost(contID uint32, idMap []specs.LinuxIDMapping) (uint32, error) {
+	if idMap == nil {
+		return contID, nil
+	}
+	for _, m := range idMap {
+		high, err := safeSum(m.ContainerID, m.Size)
+		if err != nil {
+			break
+		}
+		if contID >= m.ContainerID && contID < high {
+			hostID, err := safeSum(m.HostID, contID-m.ContainerID)
+			if err != nil || hostID == invalidID {
+				break
+			}
+			return hostID, nil
+		}
+	}
+	return invalidID, fmt.Errorf("container ID %d cannot be mapped to a host ID", contID)
+}
+
+// safeSum returns the sum of x and y. or an error if the result overflows
+func safeSum(x, y uint32) (uint32, error) {
+	z := x + y
+	if z < x || z < y {
+		return invalidID, errors.New("ID overflow")
+	}
+	return z, nil
+}
+
+func RemapDir(ctx context.Context, originalMountpoint, newMountpoint string, idMap IDMap) error {
+	idmappedSnapshotBase := filepath.Dir(newMountpoint)
+	if err := os.Mkdir(idmappedSnapshotBase, 0755); err != nil {
+		return err
+	}
+
+	// It would be preferred to use something like continuity.CopyDir, but we need to avoid copying hardlinks,
+	// unless there is a more efficient way to ensure we don't id-map the same inode multiple times
+	if err := exec.Command("cp", "-R", originalMountpoint, idmappedSnapshotBase).Run(); err != nil {
```

Comment:
copy option to be cp -a (archive option) otherwise the file permissions are not preserved.

Reply:
Opted for `cp -pR` instead as `cp -a` kept hardlinks which proved to be problematic previously when doing chown

Reply:
i am not sure this is correct, can we test the hard link case with -a. 
AFAIK replica here should not impact once we create the overlay,
we would need to test with SM with this change first also to verify it doestn break anything else.

Reply:
Ended up changing this to `cp -a --no-preserve=links`